### PR TITLE
[util] Develop SDS, SkipList, LinkedList modules

### PR DESCRIPTION
**Pull request by million-2002**
### Purpose

- Improve SDS, SkipList, LinkedList Module

### Changes

- pkg/util/sds.go pkg/util/linkedlist.go pkg/util/skiplist.go pkg/util/test_sds.go

### Review Instructions

- Read "File changed".

### CODE REVIEW

**Review by Sigma711 — CHANGES_REQUESTED** [review]
I think you should use go-fmt to specify your code format.

File: go.mod [hunk 1]
```diff
@@ -0,0 +1,3 @@
+module module1
```
Comment (Sigma711) [inline go.mod:1]:
```suggestion
module wkv
```

**Review by Sigma711 — CHANGES_REQUESTED** [review]
And you should modify it.

File: go.mod (hunk 1 above)
Comment (Sigma711) [reply go.mod]: LGTM

**Review by Sigma711 — CHANGES_REQUESTED** [review]
Change [wkv](https://github.com/InfraSail/wkv)/[.github](https://github.com/InfraSail/wkv/tree/main/.github)/[workflows](https://github.com/InfraSail/wkv/tree/main/.github/workflows)/go.yml to:
```yaml
name: Go

on:
  push:
    branches: [ main ]
  pull_request:
    branches: [ main ]

jobs:

  build:
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v2

    - name: Set up Go
      uses: actions/setup-go@v2
      with:
        go-version: 1.17

    # - name: Build
    #   run: go build -v ./...

    # - name: Test
    #   run: go test -v ./...

  golangci:
    name: lint
    runs-on: ubuntu-latest
    steps:
      - uses: actions/setup-go@v2
      - uses: actions/checkout@v2
      - name: golangci-lint
        uses: golangci/golangci-lint-action@v2
        with:
          # Optional: version of golangci-lint to use in form of v1.2 or v1.2.3 or `latest` to use the latest version
          version: v1.29

          # Optional: working directory, useful for monorepos
          # working-directory: somedir

          # Optional: golangci-lint command line arguments.
          # args: --issues-exit-code=0

          # Optional: show only new issues if it's a pull request. The default value is `false`.
          # only-new-issues: true

          # Optional: if set to true then the all caching functionality will be complete disabled,
          #           takes precedence over all other caching options.
          # skip-cache: true

          # Optional: if set to true then the action don't cache or restore ~/go/pkg.
          # skip-pkg-cache: true

          # Optional: if set to true then the action don't cache or restore ~/.cache/go-build.
          # skip-build-cache: true
```

**Comment by million-2002** [comment]
ok

**Review by Sigma711 — CHANGES_REQUESTED** [review]
git mv test/util/sds_test.go pkg/util/

File: pkg/util/skiplist.go [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+		// 初始化每层
+		for i := level - 1; i >= sl.level; i-- {
+			rank[i] = 0
+			update[i] = sl.header
+			update[i].level[i].span = sl.length
+		}
+		sl.level = level
+	}
+	// 更新节点 level，并插入新节点
+	node.SetLevel(level)
```
Comment (Sigma711) [inline pkg/util/skiplist.go:92]: `// TODO: node.SetLevel(level)`

File: pkg/util/sds.go [hunk 3]
```diff
@@ -1,110 +1,205 @@
...
 	}
-	return "OK" 
+	return "OK"
  }*/
 
- func UserAsk(ask_str string){
+func UserAsk(ask_str string) {
 	ask_split := strings.Split(ask_str, " ")
-	if ask_split[0] == "set" && ask_split[0] == "SET"{
+	if ask_split[0] == "set" && ask_split[0] == "SET" {
```
Comment (Sigma711) [inline pkg/util/sds.go:61]: I think you should change it to this style:
```go
if x := ask_split[0]; x == "set" && x == "SET" {
```
In addition, you should reconsider the operator `&&`. Does it make sense?

File: pkg/util/sds.go [hunk 4]
```diff
@@ -1,110 +1,205 @@
...
 		//存储键值对
-		SetFunc(ask_split[1],ask_split[2])
-		
-	}else if ask_split[0] == "get" && ask_split[0] == "GET"{
+		SetFunc(ask_split[1], ask_split[2])
+
+	} else if ask_split[0] == "get" && ask_split[0] == "GET" {
```
Comment (Sigma711) [inline pkg/util/sds.go:65]: I think you should change it to this style:
```go
} else if x := ask_split[0]; x == "get" && x == "GET" {
```
In addition, you should reconsider the operator `&&`. Does it make sense?

File: pkg/util/skiplist.go (hunk 2 above)
Comment (Sigma711) [inline pkg/util/skiplist.go:92]: You should change that to it:
```go
// TODO: node.SetLevel(level)
```

**Review by Sigma711 — CHANGES_REQUESTED** [review]
1. Improve the code according to my requirement in one commit.
2. Fix all warnings checked by GolangCI-Lint in the next commit.
3. Finally, fix all logic bug in it in order to make it pass all unit tests and make it sense in last commit of this time.

**Review by Sigma711 — CHANGES_REQUESTED** [review]
Undo the code comments in Line 22 - Line 26 of [.github/workflows/go.yml](https://github.com/InfraSail/wkv/pull/5/files#diff-678682767f2477de3e3c546746f8568b9a1942b2c647d32331d7e774b8ff8d9f).

**Review by Sigma711 — CHANGES_REQUESTED** [review]
bullshit

**Review by Sigma711 — CHANGES_REQUESTED** [review]
This directory .vscode should be put into `.gitignore` file.

**Comment by million-2002** [comment]
> This directory .vscode should be put into `.gitignore` file.

ok

**Review by Sigma711 — CHANGES_REQUESTED** [review]
And why delete `sds` module?

**Comment by million-2002** [comment]
> And why delete `sds` module?

sth wrong in the file

**Comment by Sigma711** [comment]
> > This directory .vscode should be put into `.gitignore` file.
> 
> ok

Modify it. I think `sds` module should not be absent.

**Comment by Sigma711** [comment]
> > And why delete `sds` module?
> 
> sth wrong in the file

Modify it. I think `sds` module should not be absent.

File: pkg/util/skiplist.go [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+		// 初始化每层
+		for i := level - 1; i >= sl.level; i-- {
+			rank[i] = 0
+			update[i] = sl.header
+			update[i].level[i].span = sl.length
+		}
+		sl.level = level
+	}
+	// 更新节点 level，并插入新节点
+	//TODO:node.SetLevel(level)
```
Comment (Sigma711) [inline pkg/util/skiplist.go:92]: 1 space after `//`
1 space after `:`

**Review by Sigma711 — CHANGES_REQUESTED** [review]
Fix it.

File: pkg/util/sds.go [hunk 6]
```diff
@@ -1,110 +1,213 @@
...
+func sdsCat(s *sdshdr, cat_str string) string {
+	//var cat_buf[] byte
+	cat_or_not++
+
+	cat_buf := []byte(cat_str)
+	s.buf = append(s.buf,cat_buf...)
+
+	for i, value := range s.buf {
+		fmt.Printf("%d,%c ", i, value)
+		s.len ++
```
Comment (Sigma711) [inline pkg/util/sds.go:154]: no space before `++`

**Review by Sigma711 — CHANGES_REQUESTED** [review]
Fix it.

File: .gitignore [hunk 7]
```diff
@@ -4,12 +4,17 @@
 *.dll
 *.so
 *.dylib
+.vscode
```
Comment (Sigma711) [inline .gitignore:7]: Line 7 should be deleted.

**Review by Sigma711 — CHANGES_REQUESTED** [review]
Fix it.

**Review by million-2002 — COMMENTED** [review]
ok

**Pull request closed without merging**